### PR TITLE
[Better Transformer] make is_causal a hint and force attn_mask to be set on `is_causal=True` in F.MHA

### LINKED ISSUE #96941: `is_causal` parameter in torch.nn.TransformerEncoderLayer.forward does not work
### 🐛 Describe the bug

As the title says, `is_causal` parameter in torch.nn.TransformerEncoderLayer.forward does not work.

```python
import torch

d_model = 3
layer = torch.nn.TransformerEncoderLayer(d_model, 1, 6, batch_first=True)
layer.eval()
x = torch.randn(1, 5, d_model)
unmasked_output = layer(x)
is_causal_output = layer(x, is_causal=True)
mask = torch.nn.Transformer.generate_square_subsequent_mask(x.size(1))
masked_output = layer(x, src_mask=mask)

assert not torch.equal(unmasked_output, is_causal_output)
assert torch.equal(masked_output, is_causal_output)
```

### Versions

```
Collecting environment information...
PyTorch version: 2.0.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.2 LTS (x86_64)
GCC version: (Ubuntu 11.3.0-1ubuntu1~22.04) 11.3.0
Clang version: Could not collect
CMake version: version 3.25.0
Libc version: glibc-2.35

Python version: 3.10.6 (main, Nov 14 2022, 16:10:14) [GCC 11.3.0] (64-bit runtime)
Python platform: Linux-5.15.0-56-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA GeForce RTX 3090
GPU 1: NVIDIA GeForce RTX 3090
GPU 2: NVIDIA GeForce RTX 3090
GPU 3: NVIDIA GeForce RTX 3090
GPU 4: NVIDIA GeForce RTX 3090
GPU 5: NVIDIA GeForce RTX 3090

Nvidia driver version: 525.60.11
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.5.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.5.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          80
On-line CPU(s) list:             0-79
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Gold 6230 CPU @ 2.10GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              20
Socket(s):                       2
Stepping:                        7
CPU max MHz:                     3900.0000
CPU min MHz:                     800.0000
BogoMIPS:                        4200.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local dtherm ida arat pln pts pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       1.3 MiB (40 instances)
L1i cache:                       1.3 MiB (40 instances)
L2 cache:                        40 MiB (40 instances)
L3 cache:                        55 MiB (2 instances)
NUMA node(s):                    2
NUMA node0 CPU(s):               0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78
NUMA node1 CPU(s):               1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79
Vulnerability Itlb multihit:     KVM: Mitigation: VMX disabled
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Mmio stale data:   Mitigation; Clear CPU buffers; SMT vulnerable
Vulnerability Retbleed:          Mitigation; Enhanced IBRS
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.24.2
[pip3] pytorch-lightning==2.0.0
[pip3] torch==2.0.0
[pip3] torch-tb-profiler==0.4.1
[pip3] torchdata==0.6.0
[pip3] torchmetrics==0.11.0
[pip3] torchtext==0.15.1
[pip3] torchvision==0.15.1
[pip3] triton==2.0.0
[conda] Could not collect
```

cc @jbschlosser @bhosmer @cpuhrsch @erichan1

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary:
This fixes an issue raised in [is_causal parameter in torch.nn.TransformerEncoderLayer.forward does not work #96941](https://github.com/pytorch/pytorch/issues/96941) where results computed with is_causal do not properly reflect causal masking.

In PyTorch 2.0, Accelerated PT Transformers added the is_causal parameter to legacy nn.Transformer* and nn.MHA APIs aligned with and intended to engage the is_causal parameter of the new scaled_dot_product_attention (SDPA) operator.

At present is_causal works differently for Transformer* modules, compared to the nn.MHA and F.MHA:
* The nn.Transformer* modules treat is_causal as an optional indicator about the format of attn_mask. This is because some layers (such as the CLIP layer use the attention mask in the layer, and thus the attn_mask was a required feature.)
* Initially, nn.MHA and F.MHA were defined to align with F.SDPA in behavior: a user may specify either the attention mask, or is_causal, but not both.  It seemed to make sense at the time to align SDPA and MHA, esp since there was a larger overlap of parameters which have since changed, e.g., with the removal of need_weights from SDPA. (See below for why this makes sense.)

Unfortunately, this does not work because of how MHA was changed to handle the need_weights parameter.  When need_weights is present, we do not (any more) call SDPA because support for need_weights was removed from SDPA before the release.  The rationale is that need_weights defeats all optimization at the foundation of SDPA performance.  Having the flag might thus mislead users into thinking they get good performance and have them disappointed when they enable a legacy feature of MHA which massively degrades performance.  (They might not think anything of enabling that, because it is on by default in MHA today, which leads to more  issues.)

Since SDPA does not (no longer) support need_weights, we need to pick a separate path which implements attention using a set of discrete operations that allocates a tensor for weights.  Alas, this code path does not have support for is_causal, because attention is implemented as matmul and using the attention mask.  Thus, is_causal has no impact.  (A substantially similar situation arises with how kpm is implemented today because Nested Tensors are not supported by torch.compile() in 2.0)

This problem was masked because all uses of legacy nn.MHA (and F.MHA) come through nn.Transformer* which called self-attention (i.e., nn.MHA) only ever with the attention mask attn_mask, and never with is_causal, a missed optimization opportunit that would have been addressed in a future performance update.  

Regrettably, always calling nn.MHA with attn_mask prevented diagnosing of the issue of not having a suitable attention mask when need_weights support was dropped from SDPA and a discrete implementation of attention was added for that scenario, and for the execution path with key_padding_mask. 

We have two options to address this issue:

Solution 1: Whenever nn.MHA and F.MHA are executed with is_causal set, we internally create a causal mask at significant expense of allocating a tensor and filling it with a triangular causal matrix.  This increases memory usage, and runtime, for allocating a causal mask.  To add insult to injury, in all current (and likely future) execution scenarios, MHA is called by a model using the nn.Transformer API which already has that matrix and passes it from nn.module to nn.module.  Then the passing in of attn_mask has to be suppressed by nn.TransformerEncoderLayer, only for nn.MHA to immediately allocate the very same tensor again to satisfy the requirement to have an attention mask for the computation. (We expect new use cases to use SDPA directly.)

Solution 2: We align the behavior of nn.MHA and F.MHA with the rest of the existing nn.Transformer API, and require the attention mask to be passed into nn.MHA in addition to is_causal as an optional indicator about the nature of the attention mask rather than as an alternative to attn_mask.  Then, when we choose the code path for processing MHA with need_weights or a key_padding_mask, we have the attn_mask passed down through the nn.Transformer* hierarchy, without the added overhead of allocating an attention mask as in scenario 1.  

This PR implements solution 2 which offers better performance and in retrospect aligns MHA better with the rest of the Transformer modules as the definition of SDPA evolved into a more streamlined high-performance operator.  It ostensibly changes how is_causal works, by requiring the attention mask to be specified.  However, as described here, and as shown in the submitted issue, is_causal is not working as intended today, so it requires a change regardless.  

In that sense, a change in API does not occur per-se, as the current implementation is not working, and a change has to occur either way to resolve the submitted issue, breaking any use cases that depend on the current implementation.  Checks exist (and more can be added) that flag any scenarios where is_causal is passed as True, but no attention mask is provided, ensuring that there's not quiet change from even the faulty behavior present in 2.0.  

As  an upside, the present implementation will improve performance by addressing the passing of the is_causal flag from Transformer modules to MHA, speeding up training for these examples, e.g., finetuning BERT, RoBERTa, XLM-R models.

Differential Revision: D44245725

### CODE REVIEW

**Comment by albanD** [comment]
I'm a bit confused by this change.
- Could you add a more detailed description on what is the motivation to do this?
- Isn't this a BC-breaking change? We should follow our [bc policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy#minimizing-the-disruption-of-bc-breaking-changes).
- Also there are quite a few unrelated comments left? We should make sure to keep the code clean and follow [high level design](https://pytorch.org/docs/stable/community/design.html).

**Comment by mikekgfb** [comment]
> I'm a bit confused by this change.
> 
> * Could you add a more detailed description on what is the motivation to do this?
> * Isn't this a BC-breaking change? We should follow our [bc policy](https://github.com/pytorch/pytorch/wiki/PyTorch's-Python-Frontend-Backward-and-Forward-Compatibility-Policy#minimizing-the-disruption-of-bc-breaking-changes).

This is still exploratory / experimental, and one possible way to resolve Issue #96941. A change here is not backward breaking because the combination of parameters addressed here currently doesn't work in the API that is being changed here, as per the issue.

**Comment by albanD** [comment]
> This is still exploratory / experimental

Aren't `nn.Transformer` and associate stable features?

**Comment by mikekgfb** [comment]
> > This is still exploratory / experimental
> 
> Aren't `nn.Transformer` and associate stable features?

The reference was to this PR, not to the API.

File: torch/nn/functional.py [hunk 1]
```diff
@@ -5182,6 +5196,29 @@ def multi_head_attention_forward(
...
+            # 3 - convert input to nested tensor to pake in kpm
+
+            # Option (1) for now.  Not sure I have the right dimensions
+            # FIXME:@drisspg
+            # if is_causal and attn_mask is None:
+            #     L = query.size(-2)
+            #     S = key.size(-2)
+            #     attn_mask = torch.triu(torch.full((L, S), float('-inf'), device=query.device, dtype=query.dtype), diagonal=1)
+
+            assert not (is_causal and attn_mask is None), "FIXME: instantiate causal attn_mask"
```
Comment (albanD) [inline torch/nn/functional.py:5215]: If this is an error that the user will see, we should have a nicer error message.

File: torch/nn/functional.py [hunk 2]
```diff
@@ -5301,6 +5338,14 @@ def multi_head_attention_forward(
     if need_weights:
         B, Nt, E = q.shape
         q_scaled = q / math.sqrt(E)
+
+        # if is_causal and attn_mask is None:
```
Comment (albanD) [inline torch/nn/functional.py:5342]: debug code?

File: test/test_transformers.py [hunk 3]
```diff
@@ -818,89 +832,6 @@ def test_bias_is_none(self):
         model(x, x, x)
         # completes without error
 
-    @parametrize("device", device_list)
```
Comment (albanD) [inline test/test_transformers.py:821]: Why remove these tests?
Things that used to be errors/unexpected need to be updated but the bulk of the test still makes sense no? Or this is already done somewhere else?

File: torch/nn/functional.py (hunk 1 above)
Comment (mikekgfb) [reply torch/nn/functional.py:5215]: The user should not.  This was a debug check.  We have two options - we can either require attn_mask to always be supplied, or instantiate it in cases where we need.  

I settled on conceptually using is_)causal as a flag about the nature of attn_mask, and always providing attn_mask is conceptually simpler.  (Basically, from a user perspective, you can specify is_causal to get better perf, or just keep using the legacy API.)

File: torch/nn/functional.py (hunk 2 above)
Comment (mikekgfb) [reply torch/nn/functional.py:5342]: yes.

File: torch/nn/functional.py [hunk 4]
```diff
@@ -5066,10 +5066,17 @@ def multi_head_attention_forward(
...
+            Default: `True`
+            Note: `needs_weight` defaults to `True`, but should be set to `False`
+            For best performance when attention weights are not nedeeded.
+            *Setting needs_weights to `True`
+            leads to a significant performance degradation.*
         attn_mask: 2D or 3D mask that prevents attention to certain positions. A 2D mask will be broadcasted for all
             the batches while a 3D mask allows to specify a different mask for the entries of each batch.
         is_causal: If specified, applies a causal mask as attention mask, and ignores
             attn_mask for computing scaled dot product attention.
+            *Warning*: If is_causal is set to True and a non-causal attention
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/functional.py:5078]: Good way to frame this.  The "Warning" was when you claim you have a causal mask in attn_mask using is_causal, but it is not.

File: torch/nn/functional.py [hunk 5]
```diff
@@ -5170,8 +5177,15 @@ def multi_head_attention_forward(
         target_type=query.dtype
     )
 
-    if is_causal:
-        attn_mask = None
+    if is_causal and attn_mask is None:
+        raise RuntimeError("Need attn mask if specifying is_causal")
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/functional.py:5181]: Good point!m We actually have a static method as a member of transformer class - generate_square_subsequent_mask L 152 transformer.py

File: torch/nn/modules/activation.py [hunk 6]
```diff
@@ -1097,8 +1097,6 @@ def forward(
         .. note::
             `batch_first` argument is ignored for unbatched inputs.
         """
-        if attn_mask is not None and is_causal:
```
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/modules/activation.py:1100]: good catch.

File: test/test_transformers.py (hunk 3 above)
Comment (mikekgfb) [reply test/test_transformers.py:821]: the removed checks were doing argument checks on combinations of inputs.  Removed them initially to not repeatedly have to change them as I explore solutions.  (github was the quickest way to run all tests, so I submitted PR revisions on every rev I wanted to test....)

In retrospect I wonder whether it's worthwhile to run all the parameter combination tests.

Comment (albanD) [reply test/test_transformers.py:821]: Testing all combination of parameters is definitely the way to go to make sure we don't regress on corner cases.
In the grand scheme of things, we should have ModuleInfo testing for this that will allow us to do this easily but I think this is fine for now.

Adding back the updated tests here sounds good.

**Review by albanD — APPROVED** [review]
SGTM Just needs a rebase!

**Comment by mxkrn** [comment]
I understand the clarification, but it would be great if the documentation was updated accordingly - currently it still says that option 1 is the current behavior.

Docs from latest `main` (https://pytorch.org/docs/main/generated/torch.nn.TransformerEncoder.html?highlight=transformerencoder#torch.nn.TransformerEncoder) say:

> is_causal ([Optional](https://docs.python.org/3/library/typing.html#typing.Optional)[[bool](https://docs.python.org/3/library/functions.html#bool)]) – If specified, applies a causal mask as mask (optional) and ignores attn_mask for computing scaled dot product attention. Default: False.

**Comment by mikekgfb** [comment]
Docs are extracted from doc strings when a new version is released. Please check the doc strings in the diff whether you consider them clear in this respect.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
